### PR TITLE
Docs for SNIP-23: Optional `recipient_code_hash` parameter

### PR DESCRIPTION
**Pull request by assafmo**
https://github.com/enigmampc/snip20-reference-impl/pull/20

### CODE REVIEW

File: SNIP-23.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# SNIP-23 - Improved UX to [SNIP-20](/SNIP-20.md) Send Operations
+
+This document describes an optional enhancement to the SNIP-20 spec. Contracts that support the existing SNIP-20 standard are still considered compliant, and clients should be written to benefit from these features when available, but provide fallback for when these features are not available.
+
+The feature specified in this document is an improved UX for the `Send` and `SendFrom` operations, aimed at enriching SNIP20 tokens composability and usage.
+
+- [SNIP-23 - Improved UX to SNIP-20 Send Operations](#snip-23---improved-ux-to-snip-20-send-operations)
+  - [The optional `recipient_code_hash` parameter](#the-optional-recipient_code_hash-parameter)
+    - [Updated `Send` interface](#updated-send-interface)
+    - [Updated `SendFrom` interface](#updated-sendfrom-interface)
```
Comment (reuvenpo) [inline SNIP-23.md:10]: I'd remove the link back to the head of the page and leave just the other three links with the indentation you used.

File: SNIP-23.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+## The optional `recipient_code_hash` parameter
+
+Up until now, any contract wishing to be invoked (notified) upon receiving a SNIP20 token, had to "register" its `code_hash` with that SNIP20 token. The reason is that on Secret Network due to security concerns, any caller to a contract must also know the SHA256 value of the contract's binary code. When a user sends tokens to a contract they usually don't know the contract's SHA256 value, and without it the SNIP20 contract wouldn't be able to invoke (notify) the receiving contract. Therefore, the SNIP20 spec made every contract wishing to receive a call upon receiving funds, to first register with that SNIP20 token.
+
+In practice, getting the SHA256 value of a contract from the chain is very easy, and implementing a contract that receives SNIP20 token transfers became a bit cumbersome, because of the added boilerplate code and the need to know in advance which tokens the contract should support.
+
+SNIP23 introduces an optional `recipient_code_hash` parameter in the `Send` and `SendFrom` operations, thus contracts are not required to "register" with the SNIP23 contract before receiving tokens.
```
Comment (reuvenpo) [inline SNIP-23.md:18]: I would add the edge cases here:

>  SNIP23 introduces an optional `recipient_code_hash` parameter in the `Send` and `SendFrom` operations. When `recipient_code_hash` is supplied, it takes precedent over the `recipient_code_hash` registered with the `RegisterReceiver` interface. This means that contracts are not required to "register" with the SNIP23 contract before receiving tokens. Sending a message without `recipient_code_hash` to an unregistered contract continues to behave the same.

BTW Assaf did we decide we want an error in that last case, or should we leave it as the "just behave like `Transfer`"?

File: SNIP-23.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+### Updated `Send` interface
+
+| Name                | Type             | Description                                                                                           | optional |
+| ------------------- | ---------------- | ----------------------------------------------------------------------------------------------------- | -------- |
+| recipient           | string           | Accounts SHOULD be a valid bech32 address, but contracts may use a different naming scheme as well    | no       |
+| recipient_code_hash | string           | A lower-case SHA256 hash of the recipient contract's code. Must have 64 characters (No leading "0x"). | yes      |
```
Comment (reuvenpo) [inline SNIP-23.md:25]: I'd phrase it as:
The SHA256 has of the recipient contract's code, encoded in lowercase hexadecimal with no leading "0x". (64 ASCII characters)

**Review by reuvenpo — CHANGES_REQUESTED** [review]
Assaf thanks for writing down this specification! 🙏🏻 
It looks good, i just added a couple of notes to clarify intention and rationale.

File: SNIP-23.md (hunk 1 above)
Comment (assafmo) [reply SNIP-23.md:10]: It's how my .md extension does table of contents. :shrug:

File: SNIP-23.md (hunk 2 above)
Comment (assafmo) [reply SNIP-23.md:18]: I wanted an error, but it breaks backwards compatibility.

File: SNIP-23.md (hunk 1 above)
Comment (reuvenpo) [reply SNIP-23.md:10]: Sure, but does it also force you to keep the ToC this way?
It's just a stylistic choice by me - people know what document they're on so having the main heading in the ToC is redundant

File: SNIP-23.md (hunk 2 above)
Comment (reuvenpo) [reply SNIP-23.md]: yeah i wanted an error too... we'll live with it i guess

File: SNIP-23.md (hunk 1 above)
Comment (assafmo) [reply SNIP-23.md]: Done

**Review by reuvenpo — APPROVED** [review]

**Pull request merged**